### PR TITLE
feat: added KNN/palette-based coloring of basic shapes w/ uninitialized color paths, background-color-based coloring of text w/ uninitialized color paths

### PR DESCRIPTION
**Pull request by heleaf**
# Description
Screen Shot 2021-06-21 at 9 40 10 PM

Related issue/PR: #78 

Change summary:
- Basic shapes (circles, ellipses, rectangles, squares, callouts) with uninitialized color paths (i.e. not set by the user) are assigned contrasting colors based on their spatial proximity to each other, using pre-created contrasting color palettes.
- Text objects with uninitialized color paths are assigned either black or white as a color, based on the color of the most immediate shape they are drawn on top of.

Motivation:
- Proximity-based color assignment and use of pre-existing, contrasting palettes offers support for users with color vision deficiency.
- Overall improvement of previous method of color sampling (selecting random r,g,b values for uninitialized color paths).
- Lack of visual clarity when previously all-black default text was layered on top of shapes with dark colors.

Dependencies: [colormap](https://github.com/bpostlethwaite/colormap#readme) module

# Implementation strategy and design decisions

Include a high-level summary of the implementation strategy and list important design decisions made, if any.
- All functions can be found in `Color.ts` and `ColorData.ts`, which are in `rendering`. `Color` stores all main functions and types, while `ColorData.ts` stores some default palettes, as well as an auxiliary function for choosing one of the default palettes randomly. 
- The hook can be found on line 67 of `index.ts`, in the function `stepUntilConvergence`.
- Colors are assigned to basic shape objects w/ uninitialized color paths, then colors are assigned to text objects with uninitialized color paths. This is calculated exactly once per `stepUntilConvergence` call (i.e. a post-processing calculation). 

### Method for coloring basic shapes (circles, ellipses, rectangles, squares, callouts)
1. Basic shapes with uninitialized color paths are filtered out from `state.shapes`
2. The list of basic shapes is used to create a k-nearest neighbor adjacency list, where the k-nearest neighbors to each node (shape) are stored for each node. Note: k is currently set to 3. This can be adjusted as necessary in the function `getNewlyColoredState`, in `Color.ts`.
3. Using a uniform sampling of one of the default palettes, the nodes (shapes) are assigned colors in a greedy fashion; colors are assigned randomly, but each shape is guaranteed to have a different color than each of its nearest neighbors. 
4. A new state is created with the updated colors for these basic shapes.

### Method for coloring text 
1. Text objects with uninitialized color paths are filtered out from `state.shapes`
2. The topmost shape in which each text object is contained is determined using basic collision detection (seeing if the centerpoint of each text object is contained in one of the shape objects).
3. The RGBA color of the topmost shape is converted to HSVA. Using thresholds for the A (alpha) and V (value) values of the shape color, the text color is assigned to be black or white. As a default, if the text centerpoint is not contained within any shape, the text will be assigned black as its color, to contrast against a white canvas. The thresholds for A and V can be adjusted in the function `createTextColorList`, in `Color.ts`.
4. A new state is created with the updated colors for these text objects.

# Examples with steps to reproduce them
- run `yarn start` from root
- in a new terminal window, navigate to `examples/set-theory-domain` and run `npx roger watch tree.sub venn.sty setTheory.dsl`
- open `localhost` in your browser and press `step to convergence` or turn on `autostep`

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `yarn test`
- [x] I ran `yarn docs` and there were no errors when generating the HTML site
- [x] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

# Open questions

Questions that require more discussion or to be addressed in future development:

### Shape color assignment
- The k-nearest neighbors for each shape is determined only using the distance between the centerpoints of each shape. Ideally, this can be improved with precise collision detection and/or bounding boxes for different shapes. 
- Color assignment is not supported for complex shapes with no "center" component (like freeform polygon, line, arrow, etc.). 
- The random (greedy) color assignment used sometimes results in close proximity shapes having "similar" colors, as the colors they are assigned are still considered "different" by the current method I have (see **B** and **E**, **C** and **G** in the image below). This might be able to be improved by calculating the minimum number of colors needed to greedily color the KNN graph (given k), and/or optimizing color selection more cleverly than the mostly random assignment.
Screen Shot 2021-06-21 at 9 27 38 PM

- The color assignment currently ignores the colors of the shapes whose colors have already been assigned. I'm a bit unsure about how to integrate the shapes whose colors have already been set by the user into the overall method. Additionally, should the user-assigned colors of various shapes also be taken into consideration for the palette used for the uninitialized color paths, or is the current method (choosing a random palette) still fine?
- The alpha (A in RGBA) for the colors assigned by this method is currently arbitrarily chosen as 0.5 (the same value it had in the colorSampler function). Perhaps this alpha should instead depend on the alpha value(s) chosen by the user for other color values they may have already assigned?
- Current default color palettes are "continuous colormaps", which may not be the best for certain types of "categorical" diagrams. 
- In the long run, exposing a palette picking option for the user at the style program level would probably be nice, so they can pick a general palette to use for their diagram. 

### Text color assignment
- Text color assignment (black or white) only takes into consideration the color of the topmost shape that each text object is drawn on top of. If transparent shapes are used, this can lead to lack of visual clarity (ex. if a lightly colored shape is layered on top of a darkly colored shape, the text color for the lightly colored shape may be chosen to be black, and appear unclear to the user, as with the label **E** in the image below). A method of detecting the "true" color that appears to the user for transparent shapes would be nice. 
Screen Shot 2021-06-21 at 9 18 33 PM

- Text color assignment only uses the centerpoint of each text object to determine if the text is contained within the shape, and assigns a color to the entire body of text accordingly. For text of a very large font and/or longer bodies of text that may not be entirely contained within a shape, this may cause visual clarity problems later on. (Perhaps a bounding box for text is also necessary?)

- The point in shape detection used for text assignment is not supported for most complex shapes (ex. freeform polygon, line, arrow, etc.).

### CODE REVIEW

**Review by hypotext — APPROVED** [review]
The system integration looks overall good to me!

Can you verify that your code works for all existing examples on our wiki page before merging? https://github.com/penrose/penrose/wiki/Example-diagrams

Btw, I didn't review the KNN code in detail - I'll leave it up to @joshsunshine can take a look to see if it satisfies the requirements for what y'all discussed.

**Comment by hypotext** [comment]
also pinging @wodeni for review :)

File: packages/core/src/renderer/ColorData.ts
Comment (wodeni) [inline packages/core/src/renderer/ColorData.ts:25]: Perhaps try https://www.npmjs.com/package/colormap? I

File: packages/core/src/renderer/Color.ts
Comment (wodeni) [inline packages/core/src/renderer/Color.ts:219]: L220 - L236 can be done with a one-liner: `const bannedColors = nodeNbors.filter(n => n < node);`?

Comment (wodeni) [reply packages/core/src/renderer/Color.ts:219]: There are probably other instances of this pattern. Lmk if you want me to do a closer pass of coding style

**Review by wodeni — CHANGES_REQUESTED** [review]
LGTM overall! Like @hypotext said, I'll leave the details of the KNN algorithm to @joshsunshine and @heleaf to discuss. My comments are mostly about coding style. I think a quick refactoring pass may be helpful for readability and maintainability, especially for the `Color` module.

**Comment by heleaf** [comment]
Thank you @hypotext and @wodeni for your comments and feedback! I'll have these changes implemented hopefully by this afternoon 😃

File: packages/core/src/index.ts
Comment (heleaf) [reply packages/core/src/index.ts:62]: The color assignment for shapes depends on the spatial proximity of different shapes in the final optimized state (and the color assignment for text depends on the color assignment for shapes), so i think the color functions have to be called after the state converges (?)

I'm not sure why, but calling the color functions in this fashion instead:
```
export const interactiveDiagram = async (
  domainProg: string,
  subProg: string,
  styProg: string,
  node: HTMLElement
): Promise<void> => {
  const updateData = (state: State) => {
    const stepped = colorUninitText(colorUninitShapes(stepUntilConvergence(state)));
    node.replaceChild(
      RenderInteractive(stepped, updateData),
      node.firstChild as Node
    );
  };
  const res = compileTrio(domainProg, subProg, styProg);
  if (res.isOk()) {
    const state: State = await prepareState(res.value);
    const optimized = colorUninitText(colorUninitShapes(stepUntilConvergence(state)));
    node.appendChild(RenderInteractive(optimized, updateData));
  } else {
    throw Error(
      `Error when generating Penrose diagram: ${showError(res.error)}`
    );
  }
};
```
doesn't seem to record the newly assigned colors in the state (and the browser ui lags when I try to step to convergence).

tl;dr I'll leave the color functions in `stepToConvergence` for now, but I'll play around to see if I can integrate them more cleanly into the overall system @hypotext

Comment (hypotext) [reply packages/core/src/index.ts:62]: > The color assignment for shapes depends on the spatial proximity of different shapes in the final optimized state (and the color assignment for text depends on the color assignment for shapes)

Oh right, I didn't realize this. Hmm, putting the function at this location seems like it might clash with the results if the user tries to optimize colors in style - the colors will be left unassigned and optimized, and then the colors will get set again in your pass, resulting in unexpected results. 

But then, your function depends on the optimized results.

Maybe your functionality can be optionally turned on via a flag in Style? e.g. 

`settings {
    ASSIGN_COLORS = true
}
`

would trigger it?

Comment (wodeni) [reply packages/core/src/index.ts:62]: > Oh right, I didn't realize this. Hmm, putting the function at this location seems like it might clash with the results if the user tries to optimize colors in style - the colors will be left unassigned and optimized, and then the colors will get set again in your pass, resulting in unexpected results.

If someone tries to optimize `A.shape.color`, for instance, will `A.shape.color` still end up in `uninitializedPaths`? If not, there's nothing to worry about. I'm actually not sure about this.

> `settings { ASSIGN_COLORS = true } `

I'd be careful about adding magic values in Style.

Comment (hypotext) [reply packages/core/src/index.ts:62]: > If someone tries to optimize A.shape.color, for instance, will A.shape.color still end up in uninitializedPaths? If not, there's nothing to worry about. I'm actually not sure about this.

Yes, because usually colors are optimized by leaving them uninitialized in Style. Since a color is not a `Float` type, it will end up in `uninitializedPaths`. The only way to really know if something is optimized is to see if it or its constituent parts appear somewhere in the energy graph, which is pretty complicated.

> I'd be careful about adding magic values in Style.

Rather than a magic value, I would think of it as a user setting, since some users might not want it. So maybe it can be a tickbox in the inspector instead? Anyway, open to other proposed solutions here.

**Comment by hypotext** [comment]
@heleaf Can you post screenshots of the results when you run your code on the other examples on our wiki page? Just curious about what those colors look like. (I think such screenshots are automatically generated by @wodeni's automator somewhere...)

File: packages/core/src/renderer/Color.ts
Comment (heleaf) [reply packages/core/src/renderer/Color.ts:457]: `prettyPrintPath` gives me the full property path for `IPropertyPath` objects (ex. `H.icon.color`) where I only need the field path (ex. `H.icon`), so could I keep this as is for now? (I think it's cleaner than using `prettyPrintPath` and then slicing the `.color` substrings out)

**Comment by heleaf** [comment]
@hypotext I have some generated diagrams uploaded (via Nimo's automator [here](https://drive.google.com/drive/folders/1agl1A2vBIp7M9Q__Si9B_9bYEI66pVbf?usp=sharing), and manually in the browser ui [here](https://drive.google.com/drive/folders/1YZTbeQpGQH_8Lr6qZjyd2jKyEouMPt0P?usp=sharing)). You might have to download the whole folder(s) and view them in finder because google drive sucks for previewing svgs :( 

Out of all the tests on the wiki page, the Cross-domain mesh-set example, the Geometry (partial) example, and the Graphs example were very slow to converge (not sure if that's already a problem).

I'm also not sure where the `graph-domain/disjoint-rects-large-canvas.sty` and `graph-domain/disjoint-rects-small-canvas.sty` files are (they weren't in my `graph-domain` folder, and I couldn't find them by calling git pull), so I haven't yet tested the 'Disjoint rectangles, contained within a large canvas (1600x1200)' and 'Disjoint rectangles, contained within a small canvas (400x300)' tests from the wiki.

**Comment by wodeni** [comment]
Looking at the diff again. A little confused by all the changes to `CHANGELOG`. Seems like this branch still needs to pull from `main` to resolve the conflicts?

**Comment by heleaf** [comment]
WIth the refactored `stepToConvergence`, I'm not sure where putting the coloring functions will be best (still in `stepToConvergence` as before, or somewhere else). The most recent commit has them in `Optimizer.ts` in the `step` function. This does assign the colors correctly when `stepToConvergence` is pressed in the UI, but it also calculates the colors at every single step (which can potentially be pretty expensive (O n^2) for diagrams with n shapes , vs. just calculating the colors once at the end). @wodeni @hypotext

File: packages/core/src/index.ts
Comment (hypotext) [reply packages/core/src/index.ts]: After thinking more about this @wodeni, I think a Style writer actually cannot optimize a color that's left uninitialized*. That means that if the color's path appears in `uninitializedPaths`, it can be set as @heleaf is doing.

*this is because if a shape's color is not set, then by default, we **don't** make each of the color components into varying floats; they can only be constants in the optimization. If someone wants to optimize a color, they have to manually make a varying path for each component, e.g. `A.r = ?; A.g = ?; A.b = ?; A.shape = Circle { color : rgba(A.r, A.g, A.b, 1.) }`. So in a sense the invariant is due to a system decision. Whereas if we **did** make uninitialized color components into varying floats by default, then we could not rely on uninitialized colors to be not optimized.

**Comment by hypotext** [comment]
@heleaf @wodeni I replied re: the `uninitializedPaths` convo: https://github.com/penrose/penrose/pull/604#discussion_r670661421.

> This does assign the colors correctly when stepToConvergence is pressed in the UI, but it also calculates the colors at every single step (which can potentially be pretty expensive (O n^2) for diagrams with n shapes , vs. just calculating the colors once at the end).

Yeah, like you mentioned, I don't think this is the most computationally efficient way to do it! I would actually prefer if the colors are set once at the end of optimization, which is indeed fine on second thought (https://github.com/penrose/penrose/pull/604#discussion_r680285706).

> possible new approach suggested by Nimo will be to make the color optimizer available in the browser UI, possibly via a button or a drop-down menu where the user can select the palette they want to use (I can use colormap module palette names for this).

I think this is a great idea. Though I would also like all settings to be stored textually, in the Style program—is a UI setting so different from a Style flag as suggested [here](https://github.com/penrose/penrose/pull/604#discussion_r670654106)?

`settings { ASSIGN_COLORS = true }`

File: packages/core/src/index.ts
Comment (wodeni) [reply packages/core/src/index.ts]: I think the actual edge case here is: if a color component participate in a constraint/objective (e.g. `ensure equal(A.shape.color[0], 1.0)`), then (1) this component (`A.shape.color[0]`) will appear in `varyingPaths` __and__ (2) the color itself is still in `uninitializedPaths`. 

If (2) is false, I think the current approach will work just fine. Otherwise, can we change the Style compiler such that it filters out values like `A.shape.color` from `uninitializedPaths`?

**Comment by wodeni** [comment]
> I think this is a great idea. Though I would also like all settings to be stored textually, in the Style program—is a UI setting so different from a Style flag as suggested [here](https://github.com/penrose/penrose/pull/604#discussion_r670654106)?
> 
> `settings { ASSIGN_COLORS = true }`

I still feel like magic values are somewhat hacky, granted that we already have features like `canvas { width: ... }`... In terms of implementation effort, the UI option is definitely less involved, as @heleaf already has an API function for assigning colors, and it's just a matter of adding a button. On the other hand, the magic value route will require some analysis of `Translation` and perhaps some error-checking passes like the ones in #589.

**Comment by wodeni** [comment]
@heleaf: after some discussion with @hypotext, I'd say it's okay to go with the UI option and then we can merge this PR. The `uninitializedPaths`/`varyingPaths`  issues deserve another PR

**Comment by heleaf** [comment]
Palette picking in browser UI! (in a separate tab) 
https://user-images.githubusercontent.com/59301744/128222581-9fdbf260-f3ea-4d7d-bb2e-39c44b2ee1ee.mov

Some questions/issues:
- I'm unsure if this method of modifying the colors directly in the current frame is preferred over a new frame in the timeline history being created every time the palette is changed (not sure how to do the latter). 
- Currently the state with original colors is lost when the user picks a new palette. I'm not sure if it's possible to store the original color scheme in case the user wants to return to it. 
- Perhaps the palette picker should be nested under an existing tab in the inspector (like `mod`?) instead of being its own thing?

**Comment by wodeni** [comment]
> Palette picking in browser UI! (using inspector tab) @wodeni @hypotext
> https://user-images.githubusercontent.com/59301744/128222581-9fdbf260-f3ea-4d7d-bb2e-39c44b2ee1ee.mov
Looks awesome!

> * I'm unsure if this method of modifying the colors directly in the current frame is preferred over a new frame in the timeline history being created every time the palette is changed (not sure how to do the latter).

I don't really have a preference. The current way seems good to me.

> * Currently the state with original colors is lost when the user picks a new palette. I'm not sure if it's possible to store the original color scheme in case the user wants to return to it.

Maybe the previous bullet can address this? I think there's `history` in `App`'s state, and you can append to that list of `State` objects whenever the user makes a color change. 

> * Perhaps the palette picker should be nested under an existing tab in the inspector (like `mod`?) instead of being its own thing?

No preference here either.

File: packages/browser-ui/src/inspector/views/Palette.tsx
Comment (wodeni) [inline packages/browser-ui/src/inspector/views/Palette.tsx:1]: The formatting in this file seems off. Do you mind running Prettier again in VSCode to fix them?

File: packages/browser-ui/src/inspector/views/viewMap.tsx
Comment (wodeni) [inline packages/browser-ui/src/inspector/views/viewMap.tsx:14]: Fix formatting here with prettier.

File: packages/browser-ui/src/inspector/views/Palette.tsx
Comment (wodeni) [inline packages/browser-ui/src/inspector/views/Palette.tsx:12]: None of the components in the inspector actually alter `history` in `App`. If you choose to do so, you'll need to pass in a callback `appendHistory` to `Inspector`. In `App`, define `appendHistory` to do:
```ts
const appendHistory = (s: State): void => this.setState({hisotry: [...this.state.history, s]})
```

**Review by wodeni — CHANGES_REQUESTED** [review]

**Comment by wodeni** [comment]
Closing due to inactivity :(.

**Pull request closed without merging**
